service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java

```diff
@@ -109,4 +148,115 @@ private StepResult updateAiNotebook(Map<String, String> metadataToUpdate, String
+      }
+      InstancesSetMachineTypeRequest setMachineTypeRequest = new InstancesSetMachineTypeRequest();
+      setMachineTypeRequest.setMachineType(machineTypeUrl);
+      computeService
+          .instances()
+          .setMachineType(projectId, location, instanceId, setMachineTypeRequest)
+          .execute();
+    }
+
+    if (acceleratorConfig != null && acceleratorConfig.getAcceleratorType() != null) {
+      gpuTypeUrl = acceleratorConfig.getAcceleratorType();
+      if (gpuTypeUrl.lastIndexOf("/") == -1) {
+        gpuTypeUrl =
+            createBaseUrl(projectId, location)
+                + "/acceleratorTypes/"
+                + gpuTypeUrl.toLowerCase().replace("_", "-");
+        acceleratorConfig.setAcceleratorType(gpuTypeUrl);
+      }
+
+      // Set the scheduling policy so that can update GPU
+      Scheduling content = new Scheduling();
+      content.setOnHostMaintenance("TERMINATE");
+      content.setAutomaticRestart(true);
+      computeService.instances().setScheduling(projectId, location, instanceId, content).execute();
+
+      InstancesSetMachineResourcesRequest setMachineResourcesRequest =
+          new InstancesSetMachineResourcesRequest();
+      setMachineResourcesRequest.setGuestAccelerators(List.of(acceleratorConfig));
+      computeService
+          .instances()
+          .setMachineResources(projectId, location, instanceId, setMachineResourcesRequest)
+          .execute();
+    }
+
+    // Wait for the instance to finish updating
+    int retryCount = 5, retryWaitSeconds = 2;
+
+    for (int i = 0; i < retryCount; i++) {
+      Instance instanceInfo =
+          computeService.instances().get(projectId, location, instanceId).execute();
+
+      if ((machineType == null || instanceInfo.getMachineType().equals(machineTypeUrl))
+          && (acceleratorConfig == null
+              || acceleratorConfig.getAcceleratorType() == null
+              || instanceInfo
+                  .getGuestAccelerators()
+                  .get(0)
+                  .getAcceleratorType()
+                  .equals(gpuTypeUrl))) {
+        break;
+      }
+      try {
+        TimeUnit.SECONDS.sleep(retryWaitSeconds);
+      } catch (InterruptedException e) {
+
+      }
+    }
+  }
+
+  /**
+   * Directly calling the gcp api to get/update instance, requires the createComputeService, see the
+   * example in https://cloud.google.com/compute/docs/reference/rest/v1/instances/get
+   */
+  public static Compute createComputeService() throws IOException, GeneralSecurityException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+
+    GoogleCredential credential = GoogleCredential.getApplicationDefault();
+    if (credential.createScopedRequired()) {
+      credential =
+          credential.createScoped(Arrays.asList("https://www.googleapis.com/auth/cloud-platform"));
```

Comment:
can you extract this to CrlService?

Reply:
similar to https://github.com/DataBiosphere/terra-workspace-manager/blob/3bd1f03cd912ea8fe93e5678a7b97876df71f1a4/service/src/main/java/bio/terra/workspace/service/crl/CrlService.java#L237.